File: typescript/src/response-mechanics.ts
import {
  governingResponse,
  governingResponseMediaMatch,
  isJSONMediaType,
  parseMediaType,
} from "./media.js";
import { resolveDeclaration, type SchemaDeclaration } from "./resolved-declaration.js";
import { classifyOpenAPI32SequentialResponse } from "./openapi32-sequential-response.js";
import type {
  OpenAPIDocument,
  OpenAPIMediaType,
  OpenAPIOperation,
  OpenAPIParameter,
  OpenAPIResponse,
} from "./types.js";

const HTTP_TOKEN = /^[!#$%&'*+.^_`|~0-9A-Za-z-]+$/;
export const DEFAULT_OPENAPI_MAX_DELIVERY_UNIT_BYTES = 10 * 1024 * 1024;

export class OpenAPIWireMechanicsError extends Error {
  constructor(
    readonly code: "ERR_REFUSED" | "ERR_PROTOCOL" | "ERR_RESPONSE_ERROR",
    message: string = code,
  ) {
    super(message);
    this.name = "OpenAPIWireMechanicsError";
  }
}

export type OpenAPIContentCodingResult = Uint8Array | ArrayBuffer | ArrayBufferView;
export type OpenAPIContentEncoder = (
  body: Uint8Array,
) => OpenAPIContentCodingResult | Promise<OpenAPIContentCodingResult>;
export type OpenAPIContentDecoder = OpenAPIContentEncoder;
export type OpenAPICharacterEncoder = (value: string) => OpenAPIContentCodingResult;
export type OpenAPICharacterDecoder = (bytes: Uint8Array) => string;

export interface OpenAPIResponseMechanicsModel {
  document: OpenAPIDocument;
  operation: OpenAPIOperation;
  parameters: OpenAPIParameter[];
  method: string;
  emptyResponse: boolean;
  maxDeliveryUnitBytes?: number;
  responseCharacterDecodings?: ReadonlyMap<string, OpenAPICharacterDecoder>;
  unaryEventStream?: {
    actualContentTypeHeader: string;
    mediaType: string;
  };
}

export function normalizeOpenAPIContentCodings<T>(
  input: Record<string, T> | undefined,
  direction: "request" | "response",
): { codecs: ReadonlyMap<string, T>; defect?: Error } {
  const codecs = new Map<string, T>();
  for (const [authored, codec] of Object.entries(input ?? {})) {
    const token = authored.trim().toLowerCase();
    if (typeof codec !== "function") {
      return { codecs, defect: new Error(`invalid ${direction} content-coding capability ${JSON.stringify(authored)}`) };
    }
    if (codecs.has(token)) {
      return { codecs, defect: new Error(`${direction} content-coding capabilities collide at ${JSON.stringify(token)}`) };
    }
    codecs.set(token, codec);
  }
  return { codecs };
}

export async function governOpenAPIRequest(
  input: RequestInfo | URL,
  init: RequestInit | undefined,
  model: OpenAPIResponseMechanicsModel,
  codecs: ReadonlyMap<string, OpenAPIContentEncoder>,
): Promise<{ input: RequestInfo | URL; init: RequestInit | undefined }> {
  const sourceHeaders = init?.headers ?? (input instanceof Request ? input.headers : undefined);
  const headers = new Headers(sourceHeaders);
  const rawCoding = headers.get("Content-Encoding") ?? "";
  let body = init?.body ?? (input instanceof Request ? input.body : null);
  if (rawCoding !== "") {
    if (body === null) requestRefusal("request Content-Encoding cannot be supplied when no request representation is emitted");
    const governing = effectiveContentEncodingParameter(model.parameters);
    if (!governing) requestRefusal("request Content-Encoding has no effective governing Header Parameter");
    if (!schemaAdmitsHeaderValue(governing.schema, rawCoding, model.document.openapi?.startsWith("3.0") ?? true)) {
      requestRefusal("request Content-Encoding is not admitted by its governing Header Parameter");
    }
    let tokens: string[];
    try {
      tokens = parsedContentCodings(rawCoding);
    } catch {
      requestRefusal("invalid Content-Encoding field value");
    }
    let bytes = await bodyBytes(body);
    for (const token of tokens) {
      if (token === "identity") continue;
      const codec = codecs.get(token);
      if (!codec) requestRefusal(`request content-coding ${JSON.stringify(token)} is unsupported`);
      try {
        bytes = codingBytes(await codec(bytes));
      } catch {
        requestRefusal(`request content-coding ${JSON.stringify(token)} failed`);
      }
    }
    body = bytesToArrayBuffer(bytes);
    headers.delete("Content-Length");
  }
  const nextInit: RequestInit = { ...init, headers, ...(body !== null ? { body } : {}) };
  if (input instanceof Request && init === undefined) {
    return { input: new Request(input, nextInit), init: undefined };
  }
  return { input, init: nextInit };
}

export async function governOpenAPIResponse(
  response: Response,
  model: OpenAPIResponseMechanicsModel,
  codecs: ReadonlyMap<string, OpenAPIContentDecoder>,
): Promise<Response> {
  const governing = governingResponse(model.operation, response.status);
  if (governing) requireGovernedResponseHeaders(governing.response, response.headers);

  // A native 3.2 sequential response must remain a stream: the invocation
  // layer owns framing and applies the delivery-unit limit per item. An
  // encoded representation is decoded below before it can be item-framed.
  const sourceContentType = response.headers.get("Content-Type");
  const sourceCoding = response.headers.get("Content-Encoding") ?? "";
  if (
    model.method.toLowerCase() !== "head"
    && response.body
    && sourceContentType !== null
    && sourceCoding === ""
    && governing
  ) {
    // THIS PRE-CHECK MAY NOT REFUSE, and Round R2's F2 finding is why. It runs
    // before a single byte has been read, so it cannot yet know whether the
    // response carries any content at all -- and §9.6 says an EMPTY response
    // "has zero content octets" and "emit[s] no operation output value", so an
    // empty body contradicts no media declaration whatever `Content-Type` the
    // peer stamped on it. (`new Response("")` stamps `text/plain;charset=UTF-8`
    // by itself, so this is the ordinary case rather than a contrived one.)
    // Refusing here made an empty 2xx an ERR_PROTOCOL on this lane while Go's
    // 3.2 lane and both engines on 3.0/3.1 completed it -- the one cell in the
    // eight-shape matrix where the two engines still disagreed after Round R.
    //
    // A media mismatch therefore FALLS THROUGH to the byte-reading path below,
    // which knows the body's length: it reports exactly this error for a
    // NON-EMPTY body and returns absence for an empty one. Nothing is weakened;
    // the same question is asked where it can be answered.
    let match: ReturnType<typeof governingResponseMediaMatch> | undefined;
    try {
      match = governingResponseMediaMatch(governing.response, sourceContentType, true, true);
    } catch {
      match = undefined;
    }
    let sequential;
    if (match) {
      try {
        sequential = classifyOpenAPI32SequentialResponse(sourceContentType, match.media);
      } catch {
        responseError("response itemSchema has no incorporated sequential framing");
      }
    }
    // `sequential` is only ever set inside the `match` guard above, so a
    // sequential classification always has the declaration it was derived from.
    const eventStream = match && parseMediaType(sourceContentType, true).base === "text/event-stream";
    if (eventStream || (model.document.openapi === "3.2.0" && sequential && match)) {
      if (match && sequential === "sse") {
        validateResponseMediaLane(match.media, sourceContentType, false, model.responseCharacterDecodings);
      }
      return response;
    }
  }

  let bytes: Uint8Array;
  const deliveryLimit = resolveOpenAPIDeliveryUnitLimit(model.maxDeliveryUnitBytes);
  try {
    bytes = await readResponseBytes(response, deliveryLimit);
  } catch (error: unknown) {
    if (error instanceof OpenAPIWireMechanicsError) throw error;
    throw new OpenAPIWireMechanicsError("ERR_PROTOCOL");
  }
  if (model.method.toLowerCase() === "head") bytes = new Uint8Array();
  const headers = new Headers(response.headers);
  const rawCoding = headers.get("Content-Encoding") ?? "";
  const successful = response.status >= 200 && response.status < 300;
  const noContent = model.method.toLowerCase() === "head" || responseBodyForbidden(response.status);
  if (rawCoding !== "") {
    let tokens: string[];
    try {
      tokens = parsedContentCodings(rawCoding);
    } catch {
      responseError("invalid Content-Encoding field value");
    }
    if (governing) {
      for (const declared of responseHeaders(governing.response, "Content-Encoding")) {
        if (!headerExactStringDomainAdmits(declared.schema as SchemaDeclaration, rawCoding, model.document.openapi?.startsWith("3.0") ?? true)) {
          responseError("actual response Content-Encoding is outside a binding-understood exact Header domain");
        }
      }
    }
    if (!noContent && (governing || !successful)) {
      for (let index = tokens.length - 1; index >= 0; index -= 1) {
        const token = tokens[index]!;
        if (token === "identity") continue;
        const codec = codecs.get(token);
        if (!codec) responseError(`response content-coding ${JSON.stringify(token)} is unsupported`);
        try {
          bytes = codingBytes(await codec(bytes));
        } catch {
          responseError(`response content-coding ${JSON.stringify(token)} failed`);
        }
        if (bytes.byteLength > deliveryLimit) {
          throw new OpenAPIWireMechanicsError("ERR_RESPONSE_ERROR");
        }
      }
    }
  }

  model.emptyResponse = bytes.length === 0;
  if (bytes.length > 0) {
    if (!governing) {
      if (successful) responseError("non-empty response has no governing Response Object");
      return responseFromBytes(response, headers, bytes);
    }
    let contentType = headers.get("Content-Type") ?? "";
    if (contentType === "") {
      contentType = "application/octet-stream";
      headers.set("Content-Type", contentType);
    }
    let match: ReturnType<typeof governingResponseMediaMatch>;
    try {
      match = governingResponseMediaMatch(governing.response, contentType, true, true);
    } catch {
      if (successful) responseError("actual response media does not match its governing declaration");
      return responseFromBytes(response, headers, bytes);
    }
    if (!match) {
      if (successful) responseError("actual response media does not match its governing declaration");
      return responseFromBytes(response, headers, bytes);
    }
    try {
      validateResponseMediaLane(
        match.media,
        contentType,
        model.document.openapi?.startsWith("3.0") ?? true,
        model.responseCharacterDecodings,
      );
    } catch (error: unknown) {
      if (successful) throw error;
      return responseFromBytes(response, headers, bytes);
    }
    if (parseMediaType(contentType, true).base === "text/event-stream" && model.unaryEventStream) {
      headers.set(model.unaryEventStream.actualContentTypeHeader, contentType);
      headers.set("Content-Type", model.unaryEventStream.mediaType);
    }
  }
  headers.delete("Content-Length");
  const body = bytes.length === 0 && responseBodyForbidden(response.status)
    ? null
    : bytesToArrayBuffer(bytes);
  return new Response(body, {
    status: response.status,
    statusText: response.statusText,
    headers,
  });
}

function responseFromBytes(response: Response, headers: Headers, bytes: Uint8Array): Response {
  headers.delete("Content-Length");
  return new Response(bytes.length === 0 && responseBodyForbidden(response.status) ? null : bytesToArrayBuffer(bytes), {
    status: response.status,
    statusText: response.statusText,
    headers,
  });
}

/** Adds an adapter-private unary alias for every governed SSE declaration. */
export function prepareOpenAPIUnaryResponseView(
  operation: OpenAPIOperation,
  unaryMediaType: string,
): void {
  for (const response of Object.values(operation.responses ?? {})) {
    if (!response || typeof response !== "object") continue;
    try {
      const match = governingResponseMediaMatch(response, "text/event-stream", true, true);
      if (match) {
        response.content ??= {};
        response.content[unaryMediaType] = match.media;
      }
    } catch { /* malformed declarations remain owned by ordinary planning */ }
  }
}

function effectiveContentEncodingParameter(parameters: OpenAPIParameter[]): OpenAPIParameter | null {
  const found = parameters.filter((parameter) =>
    parameter.in === "header" && parameter.name?.toLowerCase() === "content-encoding");
  return found.length === 1 ? found[0]! : null;
}

function responseHeaders(response: OpenAPIResponse, wanted: string): Record<string, unknown>[] {
  const headers = asRecord(response.headers);
  return Object.entries(headers ?? {})
    .filter(([name, value]) => name.toLowerCase() === wanted.toLowerCase() && asRecord(value) !== null)
    .map(([, value]) => asRecord(value)!);
}

function requireGovernedResponseHeaders(response: OpenAPIResponse, actual: Headers): void {
  const headers = asRecord(response.headers);
  for (const [name, raw] of Object.entries(headers ?? {})) {
    const declaration = asRecord(raw);
    if (name.toLowerCase() === "content-type" || declaration?.required !== true) continue;
    if (!actual.has(name)) responseError(`required response header ${JSON.stringify(name)} is absent`);
  }
}

function schemaAdmitsHeaderValue(
  schema: SchemaDeclaration,
  value: string,
  oas30: boolean,
): boolean {
  const declaration = resolveDeclaration(schema, oas30);
  return !declaration.ambiguous
    && (declaration.typeless() || declaration.admitsStringAsSoleNonNullType())
    && declaration.admitsStringEnumValue(value);
}

/**
 * Response Content-Encoding uses only the binding's deliberately small,
 * decidable raw-string subset. Each non-empty string-valued enum/const domain
 * contributes conjunctively; all other Schema constraints are descriptive at
 * this wire boundary and cannot make processors invent inverse serialization.
 */
function headerExactStringDomainAdmits(
  schema: SchemaDeclaration,
  value: string,
  oas30: boolean,
): boolean {
  const declaration = resolveDeclaration(schema, oas30);
  if (declaration.ambiguous || declaration.admitsNoInstance()) return true;
  return declaration.admitsStringEnumValue(value);
}

function parsedContentCodings(raw: string): string[] {
  const members = raw.split(",");
  if (members.length === 0) throw new Error("empty Content-Encoding");
  return members.map((member) => {
    const token = member.trim().toLowerCase();
    if (!HTTP_TOKEN.test(token)) throw new Error("invalid Content-Encoding token");
    return token;
  });
}

function validateResponseMediaLane(
  media: OpenAPIMediaType,
  contentType: string,
  oas30: boolean,
  decoders?: ReadonlyMap<string, OpenAPICharacterDecoder>,
): void {
  const parsed = parseMediaType(contentType, true);
  if (isJSONMediaType(parsed.base)) return;
  const declaration = resolveDeclaration(media.schema, oas30);
  if (isCharacterDataMedia(parsed.base) && declaration.admitsStringAsSoleNonNullType()) {
    requireSupportedCharset(parsed.params.charset ?? "utf-8", decoders);
    return;
  }
  if (declaration.typeless()) return;
  if (declaration.admitsStringAsSoleNonNullType()) {
    const format = declaration.format();
    const encoding = declaration.keywordString("contentEncoding");
    if (format.conflict || encoding.conflict) responseError("response declaration has conflicting carriage annotations");
    if ((oas30 && (format.value === "binary" || format.value === "byte")) || (!oas30 && encoding.value !== "")) {
      return;
    }
  }
  responseError(`response media ${JSON.stringify(contentType)} selects no incorporated carriage lane`);
}

function isCharacterDataMedia(base: string): boolean {
  if (base.startsWith("text/")) return true;
  return base === "application/xml" || base.endsWith("+xml");
}

function requireSupportedCharset(
  charset: string,
  decoders?: ReadonlyMap<string, OpenAPICharacterDecoder>,
): void {
  if (!["utf-8", "utf8"].includes(charset.toLowerCase()) && !decoders?.has(charset.toLowerCase())) {
    responseError(`unsupported response charset ${JSON.stringify(charset)}`);
  }
}

async function bodyBytes(body: BodyInit | ReadableStream<Uint8Array> | null): Promise<Uint8Array> {
  if (body === null) return new Uint8Array();
  return new Uint8Array(await new Response(body as BodyInit).arrayBuffer());
}

async function readResponseBytes(response: Response, limit: number): Promise<Uint8Array> {
  if (!response.body) return new Uint8Array();
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let length = 0;
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      if (!value) continue;
      length += value.byteLength;
      if (length > limit) {
        await reader.cancel();
        throw new OpenAPIWireMechanicsError("ERR_RESPONSE_ERROR");
      }
      chunks.push(value);
    }
  } catch (error: unknown) {
    try { await reader.cancel(); } catch { /* best effort */ }
    throw error;
  }
  const bytes = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) {
    bytes.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return bytes;
}

function codingBytes(value: OpenAPIContentCodingResult): Uint8Array {
  if (value instanceof Uint8Array) return value;
  if (value instanceof ArrayBuffer) return new Uint8Array(value);
  return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
}

function bytesToArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  const copy = new Uint8Array(bytes.byteLength);
  copy.set(bytes);
  return copy.buffer;
}

function responseBodyForbidden(status: number): boolean {
  return status === 101 || status === 204 || status === 205 || status === 304;
}

function requestRefusal(message: string): never {
  throw new OpenAPIWireMechanicsError("ERR_REFUSED", message);
}

function responseError(_message: string): never {
  // Protocol diagnostics are deliberately structural at the abstract SDK
  // boundary; HTTP evidence and transport prose are not portable error data.
  throw new OpenAPIWireMechanicsError("ERR_PROTOCOL");
}

function resolveOpenAPIDeliveryUnitLimit(value: number | undefined): number {
  return value !== undefined && Number.isFinite(value) && value > 0
    ? value
    : DEFAULT_OPENAPI_MAX_DELIVERY_UNIT_BYTES;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

File: go/internal/runtime/response_mechanics.go
package openapiclient

import (
	"bytes"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"

	"github.com/getkin/kin-openapi/openapi3"
)

func normalizeContentEncoders(input map[string]ContentEncoder) (map[string]ContentEncoder, error) {
	result := make(map[string]ContentEncoder, len(input))
	for token, codec := range input {
		normalized := strings.ToLower(strings.TrimSpace(token))
		if codec == nil {
			return nil, fmt.Errorf("invalid request content-coding capability %q", token)
		}
		if _, duplicate := result[normalized]; duplicate {
			return nil, fmt.Errorf("request content-coding capabilities collide at %q", normalized)
		}
		result[normalized] = codec
	}
	return result, nil
}

func normalizeContentDecoders(input map[string]ContentDecoder) (map[string]ContentDecoder, error) {
	result := make(map[string]ContentDecoder, len(input))
	for token, codec := range input {
		normalized := strings.ToLower(strings.TrimSpace(token))
		if codec == nil {
			return nil, fmt.Errorf("invalid response content-coding capability %q", token)
		}
		if _, duplicate := result[normalized]; duplicate {
			return nil, fmt.Errorf("response content-coding capabilities collide at %q", normalized)
		}
		result[normalized] = codec
	}
	return result, nil
}

func parsedContentCodings(raw string) ([]string, error) {
	members, err := splitHTTPList(raw)
	if err != nil {
		return nil, err
	}
	result := make([]string, len(members))
	for index, member := range members {
		token := strings.ToLower(strings.TrimSpace(member))
		if !httpToken(token) {
			return nil, fmt.Errorf("content-coding %q is not an HTTP token", member)
		}
		result[index] = token
	}
	return result, nil
}

func applyRequestContentCodings(request *http.Request, parameters openapi3.Parameters, edition string, codecs map[string]ContentEncoder) error {
	raw := strings.Join(request.Header.Values("Content-Encoding"), ",")
	if raw == "" {
		return nil
	}
	if (request.Body == nil || request.Body == http.NoBody) && request.Header.Get("Content-Type") == "" {
		return fmt.Errorf("request Content-Encoding cannot be supplied when the invocation emits no request representation")
	}
	parameter := effectiveContentEncodingParameter(parameters)
	if parameter == nil {
		return fmt.Errorf("request Content-Encoding has no effective governing Header Parameter")
	}
	if !schemaAdmitsHeaderValue(parameter.Schema, raw, edition) {
		return fmt.Errorf("request Content-Encoding is not admitted by its governing Header Parameter")
	}
	tokens, err := parsedContentCodings(raw)
	if err != nil {
		return err
	}
	body, err := readRequestBody(request)
	if err != nil {
		return err
	}
	for _, token := range tokens {
		if token == "identity" {
			continue
		}
		codec := codecs[token]
		if codec == nil {
			return fmt.Errorf("request content-coding %q is unsupported", token)
		}
		body, err = codec(body)
		if err != nil {
			return fmt.Errorf("request content-coding %q failed: %w", token, err)
		}
	}
	replaceRequestBody(request, body)
	return nil
}

func effectiveContentEncodingParameter(parameters openapi3.Parameters) *openapi3.Parameter {
	var found *openapi3.Parameter
	for _, ref := range parameters {
		if ref == nil || ref.Value == nil || ref.Value.In != openapi3.ParameterInHeader || !strings.EqualFold(ref.Value.Name, "Content-Encoding") {
			continue
		}
		if found != nil {
			return nil
		}
		found = ref.Value
	}
	return found
}

func schemaAdmitsHeaderValue(ref *openapi3.SchemaRef, value, edition string) bool {
	if ref == nil || ref.Value == nil {
		return true
	}
	declaration := resolveDeclaration(ref.Value, strings.HasPrefix(edition, "3.0."))
	if declaration.ambiguous || (!declaration.typeless() && !declaration.admitsStringAsSoleNonNullType()) {
		return false
	}
	for _, conjunct := range declaration.conjuncts {
		if len(conjunct.Enum) == 0 {
			continue
		}
		admitted := false
		for _, candidate := range conjunct.Enum {
			if text, ok := candidate.(string); ok && text == value {
				admitted = true
				break
			}
		}
		if !admitted {
			return false
		}
	}
	return true
}

func readRequestBody(request *http.Request) ([]byte, error) {
	if request.Body == nil {
		return nil, nil
	}
	body, err := io.ReadAll(request.Body)
	_ = request.Body.Close()
	return body, err
}

func replaceRequestBody(request *http.Request, body []byte) {
	request.Body = io.NopCloser(bytes.NewReader(body))
	request.ContentLength = int64(len(body))
	request.Header.Set("Content-Length", strconv.Itoa(len(body)))
	request.GetBody = func() (io.ReadCloser, error) {
		return io.NopCloser(bytes.NewReader(body)), nil
	}
}

func applyResponseMechanics(request *http.Request, response *http.Response, document *openapi3.T, operation *openapi3.Operation, profile string, codecs map[string]ContentDecoder, bufferEventStreams bool) (*http.Response, error) {
	if response == nil {
		return response, nil
	}
	if response.StatusCode == http.StatusSwitchingProtocols {
		closeResponse(response)
		return nil, fmt.Errorf("101 Switching Protocols is outside this binding's HTTP invocation model")
	}
	governing := governingResponse(operation, response.StatusCode)
	if governing != nil {
		if err := requireGovernedResponseHeaders(governing.response, response.Header); err != nil {
			closeResponse(response)
			return nil, err
		}
	}
	rawCoding := strings.Join(response.Header.Values("Content-Encoding"), ",")
	noContent := responseHasNoContentSemantics(request.Method, response.StatusCode)
	buffer := bufferEventStreams || rawCoding != "" || noContent
	if !buffer {
		return response, nil
	}
	body, err := readResponseBody(response)
	if err != nil {
		return nil, err
	}
	if rawCoding != "" {
		if governing != nil {
			edition := ""
			if document != nil {
				edition = document.OpenAPI
			}
			if !responseContentEncodingAdmitted(governing.response, rawCoding, edition) {
				return nil, fmt.Errorf("actual response Content-Encoding is not admitted by every governing Header Object")
			}
		}
		tokens, parseErr := parsedContentCodings(rawCoding)
		if parseErr != nil {
			return nil, parseErr
		}
		// HEAD and the status codes below carry representation metadata but no
		// response content. Validate the field grammar and governing finite
		// domain, but do not require or run a content decoder for bytes that
		// HTTP says cannot exist.
		if noContent {
			if len(body) != 0 {
				return nil, fmt.Errorf("HTTP response to %s with status %d carries forbidden content", request.Method, response.StatusCode)
			}
			replaceResponseBody(response, nil)
			return response, nil
		}
		for index := len(tokens) - 1; index >= 0; index-- {
			token := tokens[index]
			if token == "identity" {
				continue
			}
			codec := codecs[token]
			if codec == nil {
				return nil, fmt.Errorf("response content-coding %q is unsupported", token)
			}
			body, parseErr = codec(body)
			if parseErr != nil {
				return nil, fmt.Errorf("response content-coding %q failed: %w", token, parseErr)
			}
		}
	}
	if noContent {
		if len(body) != 0 {
			return nil, fmt.Errorf("HTTP response to %s with status %d carries forbidden content", request.Method, response.StatusCode)
		}
		replaceResponseBody(response, nil)
		return response, nil
	}
	if len(body) > 0 {
		if governing == nil {
			return nil, fmt.Errorf("non-empty response has no governing Response Object")
		}
		contentType := response.Header.Get("Content-Type")
		if contentType == "" {
			contentType = "application/octet-stream"
			response.Header.Set("Content-Type", contentType)
		}
		match, matchErr := governingResponseMediaMatchFor(governing.response, contentType, profile)
		if matchErr != nil {
			return nil, matchErr
		}
		if laneErr := validateResponseMediaLane(document, match.media, contentType, profile); laneErr != nil {
			return nil, laneErr
		}
	}
	replaceResponseBody(response, body)
	return response, nil
}

func responseContentEncodingAdmitted(response *openapi3.Response, value, edition string) bool {
	if response == nil {
		return true
	}
	is30 := strings.HasPrefix(edition, "3.0.")
	for name, ref := range response.Headers {
		if !strings.EqualFold(name, "Content-Encoding") || ref == nil || ref.Value == nil {
			continue
		}
		domain, known := exactHeaderStringDomain(ref.Value.Schema, is30)
		if known && !domain[value] {
			return false
		}
	}
	return true
}

func responseHasNoContentSemantics(method string, status int) bool {
	return method == http.MethodHead || (status >= 100 && status < 200) || status == http.StatusNoContent || status == http.StatusResetContent || status == http.StatusNotModified
}

func requireGovernedResponseHeaders(response *openapi3.Response, actual http.Header) error {
	if response == nil {
		return nil
	}
	for name, ref := range response.Headers {
		if strings.EqualFold(name, "Content-Type") || ref == nil || ref.Value == nil || !ref.Value.Required {
			continue
		}
		if len(actual.Values(name)) == 0 {
			return fmt.Errorf("required response header %q is absent", name)
		}
	}
	return nil
}

func responseHeader(response *openapi3.Response, name string) *openapi3.Header {
	if response == nil {
		return nil
	}
	var found *openapi3.Header
	for declared, ref := range response.Headers {
		if !strings.EqualFold(declared, name) || ref == nil || ref.Value == nil {
			continue
		}
		if found != nil {
			return nil
		}
		found = ref.Value
	}
	return found
}

func readResponseBody(response *http.Response) ([]byte, error) {
	if response.Body == nil {
		return nil, nil
	}
	body, err := io.ReadAll(response.Body)
	_ = response.Body.Close()
	return body, err
}

func replaceResponseBody(response *http.Response, body []byte) {
	response.Body = io.NopCloser(bytes.NewReader(body))
	response.ContentLength = int64(len(body))
	response.Header.Set("Content-Length", strconv.Itoa(len(body)))
}

func closeResponse(response *http.Response) {
	if response != nil && response.Body != nil {
		_ = response.Body.Close()
	}
}

func validateResponseMediaLane(document *openapi3.T, media *openapi3.MediaType, contentType, profile string) error {
	parsed, err := parseRevision3MediaType(contentType)
	if err != nil {
		return err
	}
	if document != nil && strings.HasPrefix(document.OpenAPI, "3.2.") {
		kind, sequentialErr := ClassifyOpenAPI32SequentialResponse(contentType, media)
		if sequentialErr != nil {
			return sequentialErr
		}
		if kind != "" {
			if kind == OpenAPI32SequentialSSE {
				if charset, present := parsed.params["charset"]; present && !strings.EqualFold(charset, "utf-8") && !strings.EqualFold(charset, "utf8") {
					return fmt.Errorf("text/event-stream requires UTF-8, not charset %q", charset)
				}
			}
			return nil
		}
	}
	if isJSONMediaType(parsed.base) {
		return nil
	}
	if parsed.base == "text/event-stream" {
		if charset, present := parsed.params["charset"]; present && !strings.EqualFold(charset, "utf-8") && !strings.EqualFold(charset, "utf8") {
			return fmt.Errorf("text/event-stream requires UTF-8, not charset %q", charset)
		}
	}
	oas30 := document != nil && isOpenAPI30(majorMinor(document.OpenAPI))
	declaration := resolveDeclaration(mediaSchema(media), oas30)
	if isCharacterDataMedia(parsed.base) && declaration.admitsStringAsSoleNonNullType() {
		return supportedTextCharset(parsed)
	}
	if declaration.typeless() {
		return nil
	}
	if declaration.admitsStringAsSoleNonNullType() {
		format, conflict := declaration.format()
		if conflict {
			return fmt.Errorf("response declaration has conflicting format annotations")
		}
		encoding, conflict := declaration.keywordString("contentEncoding")
		if conflict {
			return fmt.Errorf("response declaration has conflicting contentEncoding annotations")
		}
		if (oas30 && (format == "binary" || format == "byte")) || (!oas30 && encoding != "") {
			return nil
		}
	}
	return fmt.Errorf("response media %q and its resolved declaration select no incorporated carriage lane", contentType)
}
